File: elixir/lib/2021/07_the_treachery_of_whales/aoc202107.ex
defmodule AOC2021.Day07 do
  @moduledoc """
  Advent of Code 2021, day 7: The Treachery of Whales
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split(",") |> Enum.map(&String.to_integer/1)
  end

  @doc """
  Solve part 1
  """
  def part1(input), do: input |> optimize_fuel(&linear_fuel/2)

  @doc """
  Solve part 2
  """
  def part2(input), do: input |> optimize_fuel(&triangular_fuel/2)

  @doc """
  Find best position in terms of fuel use to move to

  ## Example:

      iex> optimize_fuel([0, 1, 1, 2, 8], &linear_fuel/2)
      1 + 0 + 0 + 1 + 7
  """
  def optimize_fuel(crab_pos, fuel_func) do
    1..length(crab_pos)
    |> Enum.reduce_while(fuel_func.(crab_pos, 0), fn pos, best_fuel ->
      pos_fuel = fuel_func.(crab_pos, pos)
      if pos_fuel < best_fuel, do: {:cont, pos_fuel}, else: {:halt, best_fuel}
    end)
  end

  @doc """
  Calculate fuel needed when 1 step costs 1 fuel

  ## Example:

      iex> linear_fuel([0, 1, 1, 2, 8], 5)
      5 + 4 + 4 + 3 + 3
  """
  def linear_fuel(crab_pos, pos), do: crab_pos |> Enum.map(&abs(pos - &1)) |> Enum.sum()

  @doc """
  Calculate fuel needed when each step costs 1 additional fuel

  ## Example:

      iex> triangular_fuel([0, 1, 1, 2, 8], 3)
      (1 + 2 + 3) + (1 + 2) + (1 + 2) + (1) + (1 + 2 + 3 + 4 + 5)
  """
  def triangular_fuel(crab_pos, pos) do
    crab_pos |> Enum.map(&div(abs(pos - &1) * (abs(pos - &1) + 1), 2)) |> Enum.sum()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2021/07_the_treachery_of_whales/aoc202107.py
"""AoC 7, 2021: The Treachery of Whales."""

# Standard library imports
import collections
import functools
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return collections.Counter([int(pos) for pos in puzzle_input.split(",")])


def part1(data):
    """Solve part 1."""
    return optimize(data, metric=linear_fuel(data))


def part2(data):
    """Solve part 2."""
    return optimize(data, metric=triangular_fuel(data))


def optimize(crabs, metric):
    """Find the best position based on the metric function."""

    # Guess a first to-position
    pos = round(
        sum(num_crabs * pos for pos, num_crabs in crabs.items()) / crabs.total()
    )
    pos_left, pos_right = pos - 1, pos + 1
    step = -1 if metric(pos_left) < metric(pos_right) else 1

    # Move towards the global minimum
    while metric(pos + step) < metric(pos):
        pos += step

    return metric(pos)


def linear_fuel(crabs):
    """Factory for linear fuel metrics."""

    @functools.cache
    def _linear_fuel(to_pos):
        """Calculate total fuel needed to move to a given position, each step
        costs 1 fuel."""
        return sum(num_crabs * abs(to_pos - pos) for pos, num_crabs in crabs.items())

    return _linear_fuel


def triangular_fuel(crabs):
    """Factory for triangular fuel metrics."""

    @functools.cache
    def _triangular_fuel(to_pos):
        """Calculate total fuel needed to move to a given position, each step
        costs 1 extra fuel."""
        return sum(
            num_crabs * (steps := abs(to_pos - pos)) * (steps + 1) // 2
            for pos, num_crabs in crabs.items()
        )

    return _triangular_fuel


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
